generator/src/main/java/com/graphicsfuzz/generator/mutateapi/ExpressionGenerator.java

```diff
@@ -0,0 +1,298 @@
+    this.translationUnit = translationUnit;
+    this.pipelineInfo = pipelineInfo;
+    idGenerator = new IdGenerator();
+  }
+
+  private Expr generateLiteralNumber(Value value) {
+    assert BasicType.allScalarTypes().contains(value.getType());
+
+    if (value.getType() == BasicType.FLOAT) {
+      return new FloatConstantExpr(nameFromNumber(value.getData().get(0).get()));
+    }
+    if (value.getType() == BasicType.INT) {
+      return new IntConstantExpr(nameFromNumber(value.getData().get(0).get()));
+    }
+    if (value.getType() == BasicType.UINT) {
+      return new UIntConstantExpr(nameFromNumber(value.getData().get(0).get()) + "u");
+    }
+    List<Expr> args = new ArrayList<>();
+    for (int i = 0; i < value.getType().getNumElements(); i++) {
+      assert value.getType().getElementType() != value.getType();
+      args.add(generateLiteralNumber(
+          new PrimitiveValue(value.getType().getElementType(),
+              Arrays.asList(Optional.of(value.getData().get(i).get()))
+          )));
+    }
+    return new TypeConstructorExpr(value.getType().toString(), args);
+  }
+
+  public Expr generateExpr(FactManager factManager,
+                           FunctionDefinition functionDefinition,
+                           Stmt stmt,
+                           Value value,
+                           IRandom generator) {
+    while (true) {
+      switch (generator.nextInt(2)) {
+        case 0:
+          // Generate variable fact
+          return generateVariableFact(factManager,
+              functionDefinition,
+              stmt,
+              value,
+              generator);
+        case 1:
+          // Generate function fact
+          return generateFunction(factManager,
+              functionDefinition,
+              stmt,
+              value,
+              generator);
+        case 2:
+          // Retrieving and return the existing fact from the Fact manager
+          final Optional<Expr> expr = factManager.getFact(value);
+          if (expr.isPresent()) {
+            return expr.get();
+          } else {
+            continue;
+          }
+          // Generate ConstantExpr (Float, Int)
+        case 3:
+          return generateLiteralNumber(value);
+        default:
+          return null;
+      }
+    }
+  }
+
+  // TODO: Handle the cases of vectors (vec4, vec3, vec2).
+  private String genVarName(Value value) {
+    return Constants.GLF_PRIMITIVE_VAR + "_"
+        + value.getType().toString() + "_"
+        + value.getData().get(0).get().toString().replace(".", "_");
+  }
+
+  // TODO: Requires further implementation in the case of the negative number, i.e., -0.75 could be
+  // converted to _GLF_FLOAT_NEGATIVE_0_75.
+  private String nameFromNumber(Number number) {
+    return number.toString();
+  }
+
+  private String getFunctionName(Value value) {
+    return Constants.GLF_COMPUTE + "_"
+        + value.getType().toString().toUpperCase() + "_"
```

Comment:
Again, I think the "value.getData().get(0).get()).replace(...)" stuff would be better off as a toString method on PrimitiveValue.